specs/simple-serialize.md

```diff
@@ -54,37 +54,72 @@ For convenience we alias:
 
 We recursively define the `serialize` function which consumes an object `value` (of the type specified) and returns a bytestring of type `"bytes"`.
 
-*Note*: In the function definitions below (`serialize`, `hash_tree_root`, `signed_root`, etc.) objects implicitly carry their type.
+> *Note*: In the function definitions below (`serialize`, `hash_tree_root`, `signed_root`, `is_fixed_size`, `is_variable_size` etc.) objects implicitly carry their type.
 
 ### `"uintN"`
 
+A byte string of width  `N // 8` containing the little-endian encode integer.
+
 ```python
 assert N in [8, 16, 32, 64, 128, 256]
 return value.to_bytes(N // 8, "little")
 ```
 
 ### `"bool"`
 
+* The byte `\x00` **if** the value is `False`
+* The byte `\x01` **if** the value is `True`
+
 ```python
 assert value in (True, False)
 return b"\x01" if value is True else b"\x00"
 ```
 
 ### Vectors, containers, lists
 
-If `value` is fixed-size:
+The serialized representation of composite types is comprised of two binary sections.
 
-```python
-return "".join([serialize(element) for element in value])
-```
+* The first section is fixed-size for all types, containing the concatenation of
+    - The serialized representation for each of the fixed-size elements from the value.
```

Comment:
did you consider putting offsets at their respective field locations? this maintains the fixed size of the fixed-size portion, but you'll find data about each field at the "natural" location. 

further, it allows you to extend it with new fields more easily without breaking existing data: the new fixed-size fields doesn't change the location of the offsets, and in new data, the offsets jump over the new fields. a length prefix of the fixed-size section is the final piece needed, and a newer schema with more fields can still be used to read old data. this will be very handy when upgrading to eth 2.1.

another common thing and useful thing for easy decoding: alignment. tightly packed or aligned to their size? the latter is better for mmap-style access, if you're simply storing blobs somewhere. it's nice for flash memory and other resource-constrained devices. alignment to size should be very cheap for eth2 - most fields are uint64 or bytes32 and thus already aligned, unless someone accidentally puts a bool in the wrong place, and then the whole object is ruined - having it in the spec makes it easier to rely upon, even though in practice we can use it for most existing types already.

Reply:
example: for an object whose fields are `int, list, int`, you'll encode: `fixed-length, int, offset, int, list-data`

Reply:
If you read the spec you'll see it defines the encoding just as you suggest, with the offsets intermixed with the fixed data elements.

Reply:
Right, thanks. I got confused by the bullet points, taking them to mean that all offsets come after all fixed-length fields.

I do not see that the fixed-length portion is length-prefixed - can you think of any other way to handle the following situation: we want to extend `int, list, int` to `int, list, int, int` but use the new schema to read the old data. 

if there's a variable-length field in there, a trick would be to look at the first offset. if there is none however..

Reply:
There could be a convention to have a schema id as the first field for objects that may need to be changed in the future (e.g. network messages). Then one could read the first int, choose the appropriate schema for the rest of the data, and continue as normal from there.

Reply:
> can you think of any other way to handle the following situation: we want to extend `int`, `list`, `int` to `int`, `list`, `int`, `int` but use the new schema to read the old data.

My first instinct is to say this is not supported.  My understanding is that we're addressing this at the wire protocol level so that we don't need this kind of compatibility and extending the schema would mean new networking protocol commands with new schemas.

> Right, thanks. I got confused by the bullet points, taking them to mean that all offsets come after all fixed-length fields.

This is good information.  It suggests that the spec still needs some improvement to make this concept more clear.  I'll take another stab at clarifying and cleaning up language tomorrow.

Reply:
> There could be a convention to have a schema id

> at the wire protocol level 

yes, agree with @pipermerriam that's sufficient for breaking upgrades - I don't think we need anything in this part of the spec for that kind. by breaking, I mean that it's essentially a mechanism for introducing new protocols, which may or may not be similar to existing ones.

however, the introduction of a `length` field would allow forwards and backwards-compatible upgrades - upgrades where we don't need to break existing readers - incredibly useful for long-lived formats (think x86 machine code encoding level longevity). the length then works a little bit like a smart version - new clients use it to distinguish between new and old data, while old clients continue to work without needing rewrite. the cost is fairly low (4 bytes, or whatever) - it doesn't change streaming or any other properties of the encoding.

the idea is not novel in any way - it's very commonly used in `C` structs for example when creating public `API`'s with the ability to upgrade internal details in the future.

Reply:
@arnetheduck I understand that use case but I think I have a strong-ish case against it.

By dictating that the length-prefix of the first section dictates the length of the first section, there becomes a section of bytes between the end of the first section and the beginning of the second section which is undefined and allowed to be *anything*.  It's unclear to me how you would account for this in trie hash.  It means that anyone who handles the `V2` version of those objects would have to somehow know how to hash them as both versions **and** likely need to be able to reference them with either hash.

This outlines why I dislike this approach.  It reminds me of this: https://linux-audit.com/linux-history-how-dot-files-became-hidden-files/

A seemingly innocuous choice can have significant unforeseen downstream effects.  I'm :-1: but totally willing open to being convinced otherwise, though you would have to convince me of the following two things.

1. There is a compelling reason why the wire protocol level mechanisms for upgrading would not be sufficient
2. That there is a solution for the trie hashing problem which doesn't introduce required extra complexity into every SSZ library.

Reply:
1. I believe SSZ usage, if done well, will leak outside of this phase0 spec and into other areas ETH.
 * One of the reasons to use SOS that's been mentioned is to access data from contracts efficiently. These are generally hard to upgrade and I believe would benefit from this capability - the ability to keep reading certain blobs without needing to be upgraded.
  * Passive listeners and tooling that operates in a read-only fashion on the bytes they know about already degrade gracefully.
  * nodes / entities that filter and forward based on the content of some fields, that don't need access to all data

2. this is an interesting point, thanks! I think there are at least some mitigating factors - mainly that core clients that produce and verify, and above all, sign messages, will necessarily need to be upgraded to understand what they're singing - thus the bytes can generally not be _anything_ - they need to be something credible in order to be accepted by the network as a whole. How to do the hashing in case you actually need to do that, hm, that would be trickier indeed. Might not be worth the cost, like you say.